File: app/src/main/java/org/wikipedia/analytics/PageScrollFunnel.java
package org.wikipedia.analytics;

import android.support.annotation.NonNull;

import org.json.JSONObject;
import org.wikipedia.WikipediaApp;
import org.wikipedia.util.DimenUtil;
import org.wikipedia.util.ReleaseUtil;

public class PageScrollFunnel extends TimedFunnel {
    private static final String SCHEMA_NAME = "MobileWikiAppPageScroll";
    private static final int REV_ID = 14591606;

    private final int pageId;
    private int viewportHeight;
    private int pageHeight;
    private int scrollFluxDown;
    private int scrollFluxUp;
    private int maxScrollY;

    public PageScrollFunnel(WikipediaApp app, int pageId) {
        super(app, SCHEMA_NAME, REV_ID, ReleaseUtil.isProdRelease() ? Funnel.SAMPLE_LOG_100 : Funnel.SAMPLE_LOG_ALL);
        this.pageId = pageId;
    }

    @Override protected void preprocessSessionToken(@NonNull JSONObject eventData) { }

    public void onPageScrolled(int oldScrollY, int scrollY, boolean isHumanScroll) {
        if (isHumanScroll) {
            if (scrollY > oldScrollY) {
                scrollFluxDown += (scrollY - oldScrollY);
            } else {
                scrollFluxUp += (oldScrollY - scrollY);
            }
        }
        maxScrollY = Math.max(maxScrollY, scrollY);
    }

    public void setPageHeight(int height) {
        this.pageHeight = (int) (height * DimenUtil.getDensityScalar());
    }

    public void setViewportHeight(int height) {
        this.viewportHeight = height;
    }

    public void logDone() {
        log(
                "pageID", pageId,
                "pageHeight", (int) (pageHeight / DimenUtil.getDensityScalar()),
                "scrollFluxDown", (int) (scrollFluxDown / DimenUtil.getDensityScalar()),
                "scrollFluxUp", (int) (scrollFluxUp / DimenUtil.getDensityScalar()),
                "maxPercentViewed", getMaxPercentViewed()
        );
    }

    private int getMaxPercentViewed() {
        final int maxPercent = 100;
        return Math.min(maxPercent, pageHeight == 0 ? 0 : ((maxScrollY + viewportHeight) * maxPercent / pageHeight));
    }
}

File: app/src/main/java/org/wikipedia/analytics/PageScrollFunnel.kt
package org.wikipedia.analytics

import org.json.JSONObject
import org.wikipedia.WikipediaApp
import org.wikipedia.util.DimenUtil
import org.wikipedia.util.ReleaseUtil

class PageScrollFunnel(app: WikipediaApp, private val pageId: Int) :
        TimedFunnel(app, SCHEMA_NAME, REV_ID, if (ReleaseUtil.isProdRelease) SAMPLE_LOG_100 else SAMPLE_LOG_ALL) {

    private var viewportHeight = 0
    private var pageHeight = 0
    private var scrollFluxDown = 0
    private var scrollFluxUp = 0
    private var maxScrollY = 0

    private val maxPercentViewed: Int
        get() {
            val maxPercent = 100
            return maxPercent.coerceAtMost(if (pageHeight == 0) 0
            else (maxScrollY + viewportHeight) * maxPercent / pageHeight)
        }

    override fun preprocessSessionToken(eventData: JSONObject) {}

    fun onPageScrolled(oldScrollY: Int, scrollY: Int, isHumanScroll: Boolean) {
        if (isHumanScroll) {
            if (scrollY > oldScrollY) {
                scrollFluxDown += scrollY - oldScrollY
            } else {
                scrollFluxUp += oldScrollY - scrollY
            }
        }
        maxScrollY = maxScrollY.coerceAtLeast(scrollY)
    }

    fun setPageHeight(height: Int) {
        pageHeight = (height * DimenUtil.densityScalar).toInt()
    }

    fun setViewportHeight(height: Int) {
        viewportHeight = height
    }

    fun logDone() {
        log(
                "pageID", pageId,
                "pageHeight", (pageHeight / DimenUtil.densityScalar).toInt(),
                "scrollFluxDown", (scrollFluxDown / DimenUtil.densityScalar).toInt(),
                "scrollFluxUp", (scrollFluxUp / DimenUtil.densityScalar).toInt(),
                "maxPercentViewed", maxPercentViewed
        )
    }

    companion object {
        private const val SCHEMA_NAME = "MobileWikiAppPageScroll"
        private const val REV_ID = 18118723
    }
}
